LIPS/LIP-draft_on_coverage_app.md

```diff
@@ -0,0 +1,465 @@
+                                     uint256 _timeElapsed) external;
+}
+
+contract LidoStETHBurner is IBeaconReportReceiver {
+    uint256 private coverSharesBurnRequested;
+    uint256 private nonCoverSharesBurnRequested;
+    
+    uint256 private totalCoverSharesBurnt;
+    uint256 private totalNonCoverSharesBurnt;
+    
+    address public immutable LIDO;
+    address public immutable TREASURY;
+
+    event StEthBurnRequested(
+        bool indexed isCover,
+        address indexed requestedBy,
+        uint256 amount,
+        uint256 sharesAmount
+    );
+
+    event StEthBurnt(
+        bool indexed isCover,
+        uint256 amount,
+        uint256 sharesAmount
+    );
+
+    event ExcessStETHRecovered(
+        address indexed requestedBy,
+        uint256 amount,
+        uint256 sharesAmount
+    );
+
+    event ERC20Recovered(
+        address indexed requestedBy,
+        address indexed token,
+        uint256 amount
+    );
+        
+    function getCoverSharesBurnt() returns (uint256) external view {
+        return totalCoverSharesBurnt;
+    }
+    
+    function getNonCoverSharesBurnt() returns (uint256) external view {
+        return totalNonCoverSharesBurnt;
+    }
+    
+    function getExcessStEth() return (uint256) external view {
+        uint256 sharesBurnRequested = (coverSharesBurnRequested + nonCoverSharesBurnRequested);
+        uint256 totalShares = IStEth(LIDO).sharesOf(address(this));
+
+        require (totalShares >= sharesBurnRequested);
+        
+        return IStEth(LIDO).getPooledEthByShares(totalShares - sharesBurnRequested);
+    }    
+    
+    constructor(address _treasury, address _lido) 
+    {
+        require(_treasury != address(0), "TREASURY_ZERO_ADDRESS");
+        require(_lido != address(0), "LIDO_ZERO_ADDRESS");
+        
+        TREASURY = _treasury;
+        LIDO = _lido;
+    }
+    
+    function requestStEthBurn(uint256 stEth2Burn, bool isCover) external {
+        require(stEth2Burn > 0);
+        require(IStETH(LIDO).transferFrom(msg.sender, address(this), stEth2Burn));
+        
+        uint256 sharesAmount = IStETH(LIDO).getSharesByPooledEth(stEth2Burn);
+        
+        emit StEthBurnRequested(isCover, msg.sender, stEth2Burn, sharesAmount)
+
+        if (isCover) { 
+            coverSharesBurnRequested += sharesAmount;
+        } else {
+            nonCoverSharesBurnRequested += sharesAmount;
+        }
+    }
+    
+    function recoverExcessStETH() external {
+        uint256 excessStEth = getExcessStEth();
+        
+        if (excessStEth > 0) {
+            uint256 excessSharesAmount = IStETH(LIDO).getSharesByPooledEth(excessStEth);
+            
+            emit ExcessStETHRecovered(msg.sender, excessStEth, excessSharesAmount);
+
+            IStEth(LIDO).transfer(TREASURY, excessStEth);
+        }
+    }
+    
+    //don't accept ether
+    function () {
+        throw;
+    }
+   
+    function recoverERC20(address token, uint256 amount) {
```

Comment:
Good idea! Maybe we need something like this but for NFTs 😜

Reply:
Nice catch! I am on it.